Authorisation and routing refactors

1whatleytay opened the pull request:
- Some cleanup to auth.
 - Fix name matching bug.
 - Add POST /match/run to the bot to go over existing users in config. The method **is blocking** and **will take a few minutes to finish**.
 - At least tried to make the other methods in /match not block as to not timeout `axios`.
 - Modules will now be loaded if they extend `ServerController`.

1whatleytay committed:
Cleanup authentication

1whatleytay committed:
Add authentication option

1whatleytay committed:
Fix name matcher spaceless bug

1whatleytay committed:
Add POST /match/run endpoint

1whatleytay committed:
Add assembly module loading

1whatleytay committed:
Fix authorization messages after assembly rewrite

mk5med commented on TU20Bot/Configuration/Controller.cs:
```diff
@@ -0,0 +1,13 @@
+namespace TU20Bot.Configuration {
+    // Name conflicts incoming :D
+    [AttributeUsage(AttributeTargets.Class)]
+    public class Controller : Attribute {
```
Great work! This is what we need.

Please update the name to be more accurate as to it's purpose. Something similar to ASP .NET Core's [Authorize](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authorization.authorizeattribute?view=aspnetcore-5.0) attribute will work.

1whatleytay replied:
Hmm... The controller attribute must be applied to all controllers (inherited through ServerController) :D I thought the name was appropriate, although short. Any other suggestions?

mk5med commented on TU20Bot/Configuration/Controllers/AdministratorController.cs:
```diff
@@ -0,0 +1,70 @@
+            if (!validated)
+                throw HttpException.Forbidden();
+
+            var collection = server.client.database
```
To improve legibility the "collection" variable name should reflect what type of collection it represents. Please also update the naming of `collection` in the `AdministratorController#delete` method.

1whatleytay replied:
Consider it done. Sounds like there's some unwanted code duplication.

1whatleytay replied:
Ehh, do you really think reverting the name is necessary? I like to keep with short one-word names when it doesn't impact code readability. IMO it definitely leads to better looking code.

I changed the name of collection to something a bit shorter in this PR because the `user` seemed to be implied in the context and purpose of the class, generic types and surrounding variables.
<img width="599" alt="Screen Shot 2021-02-04 at 6 18 04 PM" src="https://user-images.githubusercontent.com/32211852/106967593-716ee800-6715-11eb-9f00-f4c8d429e03f.png">

mk5med replied:
Yes, updating the name would help improve legibility, especially now that the variable is in a higher scope of the file.

mk5med commented on TU20Bot/Configuration/Controllers/MatchController.cs:
```diff
@@ -333,5 +328,17 @@ private struct CellReference {
             server.config.userRoleMatches.RemoveAt(index);
         }
+        
+        [Route(HttpVerbs.Post, "/match/run")]
+        public async Task<bool> runMatch() {
+            // No AnyAsync :thinking:
+            foreach (var match in server.config.userRoleMatches) {
+                if (!await evaluateMatch(match)) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
```
Please add documentation to this method.

1whatleytay replied:
Bruh.... it's an `AnyAsync` that calls evaluateMatch. Anything in particular you want documented?

mk5med replied:
It would be great if it had a summary indicating the purpose of the method, where/when it is intended to be used, and side effects.

mk5med commented on TU20Bot/Models/AccountModel.cs:
```diff
@@ -8,36 +8,36 @@
-        public const int SALT_SIZE = 32;
-        public const int HASH_SIZE = 32;
-        private const int ITERATIONS = 100000;
+        private const int saltSize = 32;
+        private const int hashSize = 32;
+        private const int iterations = 100000;
```
To improve legibility constants should be visually differentiated from local variables. 
Since we are not using the C# style guide standard we can use one for Java as our ruleset currently uses a subset of those rules.

Google Java Style Guide: https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

1whatleytay replied:
The style we're using is here :D https://discord.com/channels/230737273350520834/736741911150198835/737076636725411913

What's confusing to me is that constants throughout the project are using this style already, these variables were an exception.

1whatleytay replied:
It's probably even better to just set up a linter so we aren't allowed to merge these kinds of mistakes at all.

mk5med replied:
In that case we will need review the style guide as a group this time. When scanning code it helps to have constants visually differentiated from local variables within classes.

1whatleytay replied:
A lot of people think that... but I'm not sure if that argument holds water.

Do you really care if the `username` in `getUser(username)` is a constant or not? Languages like Swift and Rust are widely adopted and don't make any styling suggestions for constants (they basically look and feel like any other variable). The C# `const` keyword is really just there to make compiler optimization simpler and to avoid people 
mistakenly modifying them at runtime.

I don't mind revising the style, but I simply made this change to enforce consistency. If you want to adopt a different style, feel free to open a refactor PR. This PR is not concerned with refactoring style. 🤷

mk5med replied:
Please be civil during discussions on public forums.

1whatleytay replied:
I meant for it to be phrased as a thought experiment 😅 but I can see how it can be construed as a bit passive-aggressive. Sorry...

mk5med commented on TU20Bot/Configuration/Server.cs:
```diff
@@ -13,36 +20,76 @@ public class Server : WebServer {
+            this
+                .WithModule(new ModuleGroup("/", false).WithModule(openApi))
+                .WithModule(new AuthorizationModule("/", this, authApi))
+                .WithLocalSessionManager();
```
This is a neat solution! Please add documentation and comments.

mk5med reviewed (changes requested):
Thanks @1whatleytay this looks good! The solution with reflection to load in the controllers is very neat.

Overall everything looks good. There were a couple areas where changes are needed, all focusing on legibility, styling, and increasing documentation. The code has not been tested.

Since we have moved away from using the standard C# style guide, it is wise to use a well defined standard as it will provide one source of truth for any questions and remove surprise differences when project contributors move across different projects. As our current styles are a subset of the common Java style standard for Google I recommend following it for now and we can evaluate how well it matches our needs while we progress.

Google Java Style Guide: https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

1whatleytay committed:
Remove duplication in AdministratorController

1whatleytay committed:
Add comments to new server method.

1whatleytay committed:
Rename Controller attribute to ControllerInfo

1whatleytay commented:
Pushed.

1whatleytay committed:
Add POST /match/run description

1whatleytay commented:
Pushed again.

mk5med committed:
Update variable naming in AdministratorController

administrator renamed to administratorPermissions
collection renamed to accountCollection

mk5med committed:
Rename type UserCollection to AccountCollection

mk5med committed:
Add ControllerInfo attributes to all Controllers
